scripts/include/tools.sh

```diff
@@ -0,0 +1,248 @@
+                    if [[ -n "${PINNED_TOOLS:-}" && ! -x "${TOOLS_DIR}/$(exe_name "${tool}")" && ! -x "${VENV_DIR}/bin/$(exe_name "${tool}")" ]];
+                    then
+                        status="${status} (but not installed in ${TOOLS_DIR} or ${VENV_DIR}/bin)"
+                        rc=1
+                    fi
+                    ;;
+                1|*)
+                    status="${status} (older than ${minimum})"
+                    rc=1
+                    ;;
+            esac
+        fi
+    fi
+    status="${tool} ${status}"
+
+    if [ $rc -eq 0 ]; then
+        status="$(green "${status}")"
+    else
+        status="$(red "${status}")"
+    fi
+    echo "${status}"
+    return ${rc}
+}
+
+# tool_version returns the semantic version of the installed tool. I will
+# return "internal" for tools implemented as aliases/functions, "missing"
+# for tools that cannot be found, and "installed" if the version cannot be
+# determined. It is a fatal error if the version cannot be determined for
+# a tool that defines a minimum required version.
+function tool_version {
+    local tool=$1
+    local version=""
+    local tool_type
+    local minimum_version
+
+    tool_type="$(type -t "${tool}")"
+    minimum_version="$(var_lookup "${tool}_version")"
+
+    # (Maybe) determine installed version of the tool.
+    if [ -z "${tool_type}" ]; then
+        echo "missing"
+    else
+        # Call custom tool version function, if defined.
+        if [ -n "$(type -t "${tool}_version")" ]; then
+            version="$("${tool}_version")"
+        # only call default "$tool version" command if minimum version is defined.
+        elif [[ "${tool_type}" = "file" && -n "${minimum_version}" ]]; then
+            version="$("${tool}" version)"
+        fi
+
+        # Version number must have at least a single dot.
+        if [[ "${version}" =~ [0-9]+(\.[0-9]+)+ ]]; then
+            echo "${BASH_REMATCH[0]}"
+        elif [ "${version}" = "missing" ]; then
+            echo "${version}"
+        else
+            if [ -n "${minimum_version}" ]; then
+                die "Cannot determine '${tool}' version (requires ${minimum_version})"
+            fi
+            case "${tool_type}" in
+                file)
+                    echo "installed"
+                    ;;
+                '')
+                    echo "missing"
+                    ;;
+                *)
+                    echo "internal"
+                    ;;
+            esac
+        fi
+    fi
+}
+
+function tool_install {
+    local tool=$1
+    local sha256
+    local url
+    local version
+    version="$(var_lookup "${tool}_version")"
+
+    blue "Installing ${tool}"
+
+    # Look for custom install command first (e.g. for Python module install via pip).
+    if [ -n "$(type -t "${tool}_install")" ]; then
+        eval "${tool}_install"
+        return
+    fi
+
+    # XXX (require_tools is not reentrant) require_tools file gzip sha256sum xz
```

Comment:
What does this mean?

Reply:
Good catch. I had already fixed this, but forgot to re-enable the code here. Fixed in 8512193.